### PR TITLE
speed up kl div loss

### PR DESCRIPTION
**Pull request by li-roy**
Moved kl div loss to aten.

benchmarks for 5000 iterations on input size (1000,100)

New
```
cuda:
forward [0.9736350309103727, 0.9922929517924786, 0.9694818360731006]
input requires_grad=True:
backward [0.5595634011551738, 0.558339926879853, 0.5546616851352155]
double backward [1.2445648494176567, 1.2245905152522027, 1.2349751549772918]
target requires_grad=True:
backward (new C++) [0.9489959231577814, 0.9553070571273565, 0.9556351029314101]
double backward (new C++) [1.8184774098917842, 1.8164670099504292, 1.845708406995982]

cpu:
forward (new C++) [7.892430987209082, 8.3068826389499, 7.985283812973648]
input requires_grad=True:
backward (new C++) [4.328460982069373, 4.45323242014274, 4.27946363389492]
double backward (new C++) [5.153504415880889, 4.629372010007501, 4.712803596165031]
target requires_grad=True:
backward (new C++) [3.4181493939831853, 3.3771288259886205, 3.7086612950079143]
double backward (new C++) [0.21922698011621833, 0.1858532396145165, 0.19477044604718685]
```

Old
```
cuda:
forward [3.101281268056482, 3.068499860819429, 3.0527669726870954]
input requires_grad=True:
backward [0.5650290949270129, 0.5730433077551425, 0.5588279226794839]
double backward [1.1287697306834161, 1.13834543293342, 1.1298578432761133]
target requires_grad=True:
backward [0.9470391101203859, 0.9560198178514838, 0.9750375030562282]
double backward [1.85760727385059, 1.7989214668050408, 1.788982989732176]

cpu:
forward (new C++) [12.474591840058565, 12.511441555805504, 12.666544185951352]
input requires_grad=True:
backward (new C++) [7.660991386976093, 7.449987292289734, 7.513917901087552]
double backward (new C++) [4.073225498665124, 4.264980792999268, 4.429787891916931]
target requires_grad=True:
backward (new C++) [3.448499082121998, 3.9072313378565013, 3.2433970272541046]
double backward (new C++) [2.126378359273076, 1.9045450473204255, 1.7932004742324352]
```

### CODE REVIEW

**Comment by li-roy** [comment]
updated with more complete benchmark info

File: aten/src/ATen/nn.yaml [hunk 1]
```diff
@@ -5,11 +5,6 @@
   scalar_check:
     output: reduction != Reduction::None || self_->dim() == 0
 
-- name: kl_div(Tensor self, Tensor target, int64_t reduction=Reduction::ElementwiseMean)
-  cname: DistKLDivCriterion
```
Comment (SsnL) [inline aten/src/ATen/nn.yaml]: can you delete `DistKLDivCriterion` as well?

Comment (li-roy) [reply aten/src/ATen/nn.yaml]: it's used in legacy nn :(

File: aten/src/ATen/native/Loss.cpp [hunk 2]
```diff
@@ -52,23 +51,35 @@ Tensor triplet_margin_loss(const Tensor& anchor, const Tensor& positive, const T
...
 
-  if (reduction == Reduction::ElementwiseMean) {
-    return output.sum() / output.numel();
-  } else if (reduction == Reduction::Sum) {
-    return output.sum();
-  }
-  return output;
+Tensor kl_div_loss(const Tensor& input, const Tensor& target, int64_t reduction) {
+  auto zeros = at::zeros_like(target);
+  auto output_pos = target * (at::log(target) - input);
```
[the remark this answers is not in the document]
Comment (soumith) [reply aten/src/ATen/native/Loss.cpp]: @ailzhang no, we expect `input` to already be in log-space

File: aten/src/ATen/native/Loss.cpp [hunk 3]
```diff
@@ -1,11 +1,22 @@
...
 #include "ATen/NativeFunctions.h"
+#include "ATen/Dispatch.h"
+#include "ATen/CPUApplyUtils.h"
 
 #define EPSILON 1e-12
 
 
 namespace at { namespace native {
 
+Tensor apply_loss_reduction(const Tensor& unreduced, int64_t reduction) {
```
Comment (SsnL) [inline aten/src/ATen/native/Loss.cpp]: nit: static inline this.

File: aten/src/ATen/native/Loss.cpp [hunk 4]
```diff
@@ -52,23 +51,38 @@ Tensor triplet_margin_loss(const Tensor& anchor, const Tensor& positive, const T
...
 
+Tensor kl_div(const Tensor& input, const Tensor& target, int64_t reduction) {
+  auto zeros = at::zeros_like(target);
+  auto output_pos = target * (at::log(target) - input);
+  auto output = at::where(target > 0, output_pos, zeros);
+  return apply_loss_reduction(output, reduction);
+}
+
+Tensor kl_div_backward_cpu(const Tensor& grad, const Tensor& input, const Tensor& target, int64_t reduction) {
+  auto grad_input = grad.type().zeros_like(input);
```
Comment (SsnL) [inline aten/src/ATen/native/Loss.cpp]: This is fine. But you can just use `at::zeros_like(input)` in future. :)

File: aten/src/ATen/native/Loss.cpp [hunk 5]
```diff
@@ -52,23 +51,38 @@ Tensor triplet_margin_loss(const Tensor& anchor, const Tensor& positive, const T
...
+          if (target_val > 0) {
+            grad_input_val = -target_val * grad_val;
+          }
+        });
+  });
   if (reduction == Reduction::ElementwiseMean) {
-    return output.sum() / output.numel();
-  } else if (reduction == Reduction::Sum) {
-    return output.sum();
+    return grad_input / input.numel();
```
Comment (SsnL) [inline aten/src/ATen/native/Loss.cpp]: The case for `Reduction::Sum` seems removed. I didn't check the maths, but am wondering if the code works for both `Sum` and `None` cases.

**Review by ssnl — APPROVED** [review]
Looks good. But please make sure that maths is correct before merging.

File: aten/src/ATen/native/Loss.cpp (hunk 5 above)
Comment (li-roy) [reply aten/src/ATen/native/Loss.cpp]: Yeah, that's intended. We only need to divide by numel in this case. The expand conveniently allows us to deal with the other two cases simultaneously.
